src/brod_transaction.erl

```diff
@@ -0,0 +1,412 @@
+%%%
+%%%   Copyright (c) 2023 @axs-mvd and contributors
+%%%
+%%%   Licensed under the Apache License, Version 2.0 (the "License");
+%%%   you may not use this file except in compliance with the License.
+%%%   You may obtain a copy of the License at
+%%%
+%%%       http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%%   Unless required by applicable law or agreed to in writing, software
+%%%   distributed under the License is distributed on an "AS IS" BASIS,
+%%%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%%   See the License for the specific language governing permissions and
+%%%   limitations under the License.
+%%%
+
+%% @doc A `brod_transaction' is a `gen_server' that orchestate a set of
+%% producers to store messages within a transaction, it also supports
+%% committing offsets in the same transaction.
+%%
+%% Simple produce sample:
+%%
+%%  {ok, Tx} = brod_transaction:new(Client, TxId, []),
+%%  lists:foreach(fun(Partition) ->
+%%                    Key = rand(), Value = rand(),
+%%                    {ok, _Offset} =
+%%                    brod_transaction:produce(Tx,
+%%                                             Topic,
+%%                                             Partition,
+%%                                             Key,
+%%                                             Value),
+%%                end, Partitions),
+%%  brod_transaction:commit(Tx),
+%%
+%% handle callback of a group suuscriber using offset commit within a
+%% transaction:
+%%
+%%  handle_message(Topic,
+%%                 Partition,
+%%                 #kafka_message{ offset = Offset
+%%                               , key = Key
+%%                               , value = Value},
+%%                 #{ client := Client
+%%                  , group_id := GroupId} =  State) ->
+%%    {ok, Tx} = brod_transaction:new(Client),
+%%    {ok, _ProducedOffset} = brod_transaction:produce(Tx, ?TOPIC_OUTPUT, Partition, Key, Value),
+%%    ok = brod_transaction:txn_add_offsets(Tx, GroupId, #{{Topic, Partition} => Offset}),
+%%    ok = brod_transaction:commit(Tx)
+%%
+%%    {ok, ack_no_commit, State}.
+%%
+
+-module(brod_transaction).
+-behaviour(gen_server).
+
+% gen_server callbacks
+-export([ init/1
+        , handle_cast/2
+        , handle_call/3
+        , handle_info/2
+        , terminate/2
+        , code_change/3]).
+
+% public API
+-export([ produce/5
```

Comment:
Code style: please move API callback exports and definitions to the top of the module.
Usually we keep the following structure to the modules:
```erlang
%%================================================================================
%% API funcions
%%================================================================================
Interesting stuff, i.e. API functions go first

%%================================================================================
%% gen_server callbacks:
%%================================================================================

callbacks go here

%%================================================================================
%% Internal functions
%%================================================================================

everything that is not exported is at the end
```